### DISCUSSION TITLE
Simplify and expand reachability analysis

### DISCUSSION THREAD

**Discussion by sharwell** [dotnet/csharplang#664] [General]
As pointed out by @gafter in https://github.com/dotnet/csharplang/issues/656#issuecomment-306320514, the language specification currently describes reachability separately from definite assignment. The separate definition of reachability ends up making the language specification more complex, while at the same time limiting the ability of conforming compilers to report warnings related to reachability. The underlying need for the limitation appears to be due to the following wording (emphasis mine):

> **A warning is reported** if the compiler determines that a statement is unreachable. It is specifically not an error for a statement to be unreachable.

I propose the following pair of changes to the C# language:

1. Relax the warning requirement related to reachability, as follows:

    > A warning ~~is~~ **may be** reported if the compiler determines that a statement is unreachable. It is specifically not an error for a statement to be unreachable.

2. Eliminate the current description of reachability (which crosses into the descriptions of several statements), and replace it with the following (credit to @gafter for the original text, which I have modified to account for expressions and local functions):

    > To determine whether a particular statement, expression, or end point is reachable, the compiler performs flow analysis according to the definite assignment rules in section 5.3. If a hypothetical variable *v* with a compiler-generated unique name that is not definitely assigned at the beginning of the method body is found to be definitely assigned at the beginning of a statement or expression, then that statement or expression is unreachable. Conversely, if *v* is not definitely assigned at the beginning of a statement or expression, then the statement or expression is reachable. Similarly, the end point of a statement or expression is reachable if and only if *v* is not definitely assigned after that statement or expression.
    >
    > Note that definite assignment analysis for local functions occurs in multiple locations. A statement, expression, or end point within a local function is reachable if and only if *v* is not definitely assigned at that location for at least one such point of local function definite analysis.

While simultaneously simplifying the language and increasing the power of reachability analysis in C#, the above changes allow a conforming implementation to preserve compatibility for existing code bases. Should a complementing feature (e.g. warning waves or `/features:strict`) allow for breaking behavioral changes with respect to compiler warnings, these changes would enable an implementation to provide additional information regarding unreachable statements and thus reduce the likelihood of reachability-related programming errors to enter production.

**Comment by gafter** [comment]
You don't need any special text for local functions. All of its analyses will get the same result for reach ability.

**Comment by sharwell** [comment]
@gafter I do not disagree outright. However, I do note the following issues:

1. The specification for local functions is currently incomplete, and does not accurately describe the state of the compiler as would be considered part of the last stable release.
2. Definite assignment analysis for captured variables, as *v* would be for a local function, ends up reporting warnings at the point where the local function is invoked or converted to a delegate. The intent for reachability analysis is to report warnings at a finer granularity, or in the case of an unused local function, on the declaring statement itself.

I'm hesitant to remove the reference to local functions from the text prior to reviewing an updated form of the specification (1). However, (2) suggests that my wording may need to be revised somewhat.

**Comment by paulomorgado** [comment]
@sharwell, that left my brain in a knot. Can you give examples of reachability vs. definite assignment?

**Comment by sharwell** [comment]
@paulomorgado Don't brains always look like knots?

Examples of reachability:

* Error CS0163 "Control cannot fall through from one case label to another" is reported when the end of a switch section is *reachable*
* Warning CS0162 "Unreachable code detected" is reported when, e.g., you have a `throw` statement followed by a `return` statement:

    ```csharp
    throw new NotImplementException();
    return null;
    ```

Examples of definite assignment:

* Error CS0165 "Use of unassigned local variable" is reported when you declare a local variable and then use it in a location where it might not have a value.
* Error CS0177 "The out parameter 'parameter' must be assigned to before control leaves the current method" is reported when you return from a method but haven't assigned a value to one or more `out` parameters.
* Error CS0171 "Field 'field' must be fully assigned before control is returned to the caller" is reported if you have a `struct` constructor that fails to assign a value to a field.

As it turns out, the two concepts are described separately in the language specification, but reachability is implemented as a specific subset of the functionality of definite assignment. Specifically, paths in code which would result in unreachable code afterwards have the definite assignment behavior "all variables are definitely assigned after [whatever it is]". You can see this in cases like the following:

```csharp
public void Example()
{
    int x;
    //goto SkipTheError; // <-- uncomment this line and the compile error goes away
    Console.WriteLine($"It would be an error to try and print {x}");

SkipTheError:
    return;
}
```

It happens to also be the case that definite assignment covers many special cases which reachability analysis does not. For example, reachability analysis doesn't cover many subexpressions. This can be seen in some bizarre cases like the following. Reachability analysis in the language says that both the true and false paths of the `if` statement are reachable. Definite assignment analysis, on the other hand, has assigned a value to `y` because *neither* branch is reachable per the more comprehensive definition in this proposal. If this proposal were accepted, the only remaining error (regarding fall through) would also stop getting reported. In addition, compiler authors would now be allowed (not required) to report a warning on each of the `Console.WriteLine` calls indicating the statement is unreachable.

```csharp
switch ((int)x)
{
case 1:
    int y;
    if ((bool?)null ?? throw new Exception())
    {
        Console.WriteLine(y);
    }
    else
    {
        Console.WriteLine(y);
    }

default:
    break;
}
```

**Comment by paulomorgado** [comment]
I think I got it!

So, to use definite assignment to determine reachability would be like having a phantom variable that is assigned before the statement and test for definite assignment. If it's not definitely assigned, than the statement is not reachable. Is that it?

**Comment by sharwell** [comment]
@paulomorgado Close, except the phantom variable is declared but not assigned to, and you are testing to see if it *is* definitely assigned to determine that something is not reachable.

**Comment by andy-hanson** [comment]
Do you know why only the first line here has an unreachable-code warning?
```cs
void A() { throw new Exception(); return;  }
void B() { return; throw new Exception(); }
```

**Comment by gafter** [comment]
The compiler has an exception (does not report unreachable code warnings) for the following three statement types:
- the empty statement `;`
- the block statement `{ }`
- the throw statement
